File: adaptived/tests/ftests/log.py
# Copyright (c) 2024, Oracle and/or its affiliates.
# DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
#
# This code is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 only, as
# published by the Free Software Foundation.
#
# This code is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# version 2 for more details (a copy is included in the LICENSE file that
# accompanied this code).
#
# You should have received a copy of the GNU General Public License version
# 2 along with this work; if not, write to the Free Software Foundation,
# Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
#
# Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
# or visit www.oracle.com if you need additional information or have any
# questions.

#
# Log class for the adaptived functional tests
#
# Note: based on libcgroup's ftests.py
#
#

import datetime
import consts

log_level = consts.DEFAULT_LOG_LEVEL
log_file = consts.DEFAULT_LOG_FILE
log_fd = None


class Log(object):

    @staticmethod
    def log(msg, msg_level=consts.DEFAULT_LOG_LEVEL):
        global log_level, log_file, log_fd

        if log_level >= msg_level:
            if log_fd is None:
                Log.open_logfd(log_file)

            timestamp = datetime.datetime.now().strftime('%b %d %H:%M:%S')
            log_fd.write('{}: {}\n'.format(timestamp, msg))

    @staticmethod
    def open_logfd(log_file):
        global log_fd

        log_fd = open(log_file, 'a')

    @staticmethod
    def log_critical(msg):
        Log.log('CRITICAL: {}'.format(msg), consts.LOG_CRITICAL)

    @staticmethod
    def log_warning(msg):
        Log.log('WARNING: {}'.format(msg), consts.LOG_WARNING)

    @staticmethod
    def log_debug(msg):
        Log.log('DEBUG: {}'.format(msg), consts.LOG_DEBUG)

# vim: set et ts=4 sw=4:

File: adaptived/src/log.c
/*
 * Copyright (c) 2024, Oracle and/or its affiliates.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */
/**
 * Logging for adaptived
 *
 */

#include <stdbool.h>
#include <assert.h>
#include <stdarg.h>
#include <syslog.h>
#include <systemd/sd-journal.h>

#include "adaptived-internal.h"

int log_level = LOG_ERR;
enum log_location log_loc = LOG_LOC_STDERR;

static void _log(int priority, const char *fmt, va_list *ap)
{
	switch(log_loc) {
		case LOG_LOC_JOURNAL:
			sd_journal_printv(priority, fmt, *ap);
			break;
		case LOG_LOC_SYSLOG:
			vsyslog(priority, fmt, *ap);
			break;
		case LOG_LOC_STDOUT:
			vfprintf(stdout, fmt, *ap);
			break;
		case LOG_LOC_STDERR:
			vfprintf(stderr, fmt, *ap);
			break;
		default:
			assert(true);
			break;
	}
	va_end(*ap);
}

API void adaptived_err(const char *fmt, ...)
{
	va_list ap;

	if (log_level >= LOG_ERR) {
		va_start(ap, fmt);
		_log(LOG_ERR, fmt, &ap);
		va_end(ap);
	}
}

API void adaptived_wrn(const char *fmt, ...)
{
	va_list ap;

	if (log_level >= LOG_WARNING) {
		va_start(ap, fmt);
		_log(LOG_WARNING, fmt, &ap);
		va_end(ap);
	}
}

API void adaptived_info(const char *fmt, ...)
{
	va_list ap;

	if (log_level >= LOG_INFO) {
		va_start(ap, fmt);
		_log(LOG_INFO, fmt, &ap);
		va_end(ap);
	}
}

API void adaptived_dbg(const char *fmt, ...)
{
	va_list ap;

	if (log_level >= LOG_DEBUG) {
		va_start(ap, fmt);
		_log(LOG_DEBUG, fmt, &ap);
		va_end(ap);
	}
}
